### PR TITLE
Gun

### PR DESCRIPTION
**Pull request by alex-strizhakov**
Support for ipv4 & ipv6

### CODE REVIEW

File: lib/tesla/adapter/gun.ex [hunk 1]
```diff
@@ -200,16 +206,19 @@ if Code.ensure_loaded?(:gun) do
 
       tls_opts =
         with "https" <- uri.scheme,
-             true <- opts[:certificates_verification],
-             host <- to_charlist(uri.host) do
-          host = :idna.encode(host)
+             true <- opts[:certificates_verification] do
+          formatted_host =
+            case format_host(uri.host) do
+              {:domain, domain} -> domain
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]: `domain` is charlist or string?

File: lib/tesla/adapter/gun.ex [hunk 2]
```diff
@@ -420,5 +429,17 @@ if Code.ensure_loaded?(:gun) do
...
+      charlist = to_charlist(string)
+
+      case :inet.parse_address(charlist) do
+        {:error, :einval} ->
+          {:domain, :idna.encode(charlist)}
+
+        {:ok, ip} when is_tuple(ip) and tuple_size(ip) in [4, 8] ->
+          {:ip, ip}
+      end
+    end
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]: Would it make sense to have `domain_or_ss_a_string(uri)` instead? (name TBD) 

Seems like all use cases are to get encoded domain OR fallback to `uri.host` 🤔

File: test/support/adapter_case.ex [hunk 3]
```diff
@@ -2,8 +2,14 @@ defmodule Tesla.AdapterCase do
...
-      @http "http://0.0.0.0:#{Application.get_env(:httparrot, :http_port)}"
-      @https "https://0.0.0.0:#{Application.get_env(:httparrot, :https_port)}"
+      # Needed for SSL test with certificate verification
+      if @adapter == Tesla.Adapter.Ibrowse do
+        @http "http://0.0.0.0:#{Application.get_env(:httparrot, :http_port)}"
+        @https "https://0.0.0.0:#{Application.get_env(:httparrot, :https_port)}"
+      else
+        @http "http://localhost:#{Application.get_env(:httparrot, :http_port)}"
+        @https "https://localhost:#{Application.get_env(:httparrot, :https_port)}"
+      end
```
Comment (teamon) [inline test/support/adapter_case.ex]: It would be better to override this in Ibrowse tests module

File: lib/tesla/adapter/gun.ex (hunk 1 above)
Comment (alex-strizhakov) [reply lib/tesla/adapter/gun.ex]: `domain` here must be charlist

File: lib/tesla/adapter/gun.ex (hunk 2 above)
Comment (alex-strizhakov) [reply lib/tesla/adapter/gun.ex]: I renamed `format_host` function to return domain or ip. 
Created second function to return domain or fallback for ips. As `gun` wants domain or ip tuple to connect, but in other places needed fallback for ips.

File: test/support/adapter_case.ex (hunk 3 above)
Comment (alex-strizhakov) [reply test/support/adapter_case.ex]: Done 🙂

**Pull request merged**